File: agents/skills/claude-code/canary-fail-fast/scripts/digest.py
"""Loud, categorized failure digest (pure).

Turns a list of failures into a terse CI-log digest + `::error` workflow
annotations + a non-zero exit code, so an engineer triages from the run log
without opening the HTML report.
"""

from __future__ import annotations

from dataclasses import dataclass

from failures import FAILURE_CATEGORIES, categorize_failure


@dataclass
class Digest:
    text: str
    annotations: list
    exit_code: int


def _first_line(error, limit: int = 160) -> str:
    if not error:
        return "(no error message)"
    for line in error.splitlines():
        line = line.strip()
        if line:
            return line[:limit]
    return "(no error message)"


def build_digest(failures: list) -> Digest:
    if not failures:
        return Digest(text="✅ 0 failing tests.", annotations=[], exit_code=0)

    n = len(failures)
    by_cat: dict = {}
    for f in failures:
        by_cat.setdefault(categorize_failure(f.error), []).append(f)

    lines = [f"❌ {n} failing test{'s' if n != 1 else ''} — triage by category:", ""]
    for cat in FAILURE_CATEGORIES:
        bucket = by_cat.get(cat)
        if not bucket:
            continue
        lines.append(f"  {cat} ({len(bucket)}):")
        for f in bucket:
            lines.append(f"    - {f.title} — {_first_line(f.error)}")
    text = "\n".join(lines)

    annotations: list = []
    for f in failures:
        cat = categorize_failure(f.error)
        loc = ""
        if f.file:
            loc += f"file={f.file},"
        if f.line is not None:
            loc += f"line={f.line},"
        annotations.append(
            f"::error {loc}title=Test failure::{f.title} — {cat}: {_first_line(f.error)}"
        )

    return Digest(text=text, annotations=annotations, exit_code=1)

File: agents/skills/claude-code/canary-fail-fast/scripts/digest.mjs
// digest -- loud, categorized failure digest (pure). Ported behavior-for-
// behavior from the Python original.
//
// Turns a list of failures into a terse CI-log digest + `::error` workflow
// annotations + a non-zero exit code, so an engineer triages from the run log
// without opening the HTML report.

import { FAILURE_CATEGORIES, categorizeFailure } from './failures.mjs';

// Output glyphs kept as \u escapes so the source stays ASCII while the emitted
// text is byte-identical to the Python original.
const CHECK = '\u2705'; // white heavy check mark
const CROSS = '\u274c'; // cross mark
const DASH = '\u2014'; // em dash

/**
 * @typedef {{text: string, annotations: string[], exitCode: number}} Digest
 */

function firstLine(error, limit = 160) {
  if (!error) return '(no error message)';
  for (const raw of error.split(/\r\n|\r|\n/)) {
    const line = raw.trim();
    if (line) return line.slice(0, limit);
  }
  return '(no error message)';
}

/** Build the digest text, `::error` annotations, and exit code from failures. */
export function buildDigest(failures) {
  if (!failures.length) {
    return { text: `${CHECK} 0 failing tests.`, annotations: [], exitCode: 0 };
  }

  const n = failures.length;
  const byCat = new Map();
  for (const f of failures) {
    const cat = categorizeFailure(f.error);
    if (!byCat.has(cat)) byCat.set(cat, []);
    byCat.get(cat).push(f);
  }

  const lines = [
    `${CROSS} ${n} failing test${n !== 1 ? 's' : ''} ${DASH} triage by category:`,
    '',
  ];
  for (const cat of FAILURE_CATEGORIES) {
    const bucket = byCat.get(cat);
    if (!bucket || !bucket.length) continue;
    lines.push(`  ${cat} (${bucket.length}):`);
    for (const f of bucket) {
      lines.push(`    - ${f.title} ${DASH} ${firstLine(f.error)}`);
    }
  }
  const text = lines.join('\n');

  const annotations = [];
  for (const f of failures) {
    const cat = categorizeFailure(f.error);
    let loc = '';
    if (f.file) loc += `file=${f.file},`;
    if (f.line !== null && f.line !== undefined) loc += `line=${f.line},`;
    annotations.push(
      `::error ${loc}title=Test failure::${f.title} ${DASH} ${cat}: ${firstLine(f.error)}`,
    );
  }

  return { text, annotations, exitCode: 1 };
}
